### PR TITLE
Moved subscribe and filter into network controller

### LINKED ISSUE #16573: Move subscription/filter middleware into network client
Our middleware for handling subscription and filter-related methods (`eth-json-rpc-filters`) currently lives in our RPC pipeline _ahead_ of the network stack. This is network middleware, it belongs in the network client with the rest of the network middleware.

### PR DESCRIPTION
**Pull request by BelfordZ**
Our middleware for handling subscription and filter-related methods (`eth-json-rpc-filters`) currently lives in our RPC pipeline ahead of the network stack. This commit moves this middleware to the network client middleware instead. There are two reasons for this change. First, this middleware wraps RPC methods that are supported by the network. Second, it is necessary for this middleware to live with the network client so that it will aid us in unifying the NetworkController in this repo and the NetworkController in the `controllers` repo.

---

Fixes: #16573

### CODE REVIEW

File: app/scripts/controllers/network/createJsonRpcClient.js [hunk 1]
```diff
@@ -19,11 +21,22 @@ const getTestMiddlewares = () => {
 export default function createJsonRpcClient({ rpcUrl, chainId }) {
   const fetchMiddleware = createFetchMiddleware({ rpcUrl });
   const blockProvider = providerFromMiddleware(fetchMiddleware);
+
+  // create subscription polyfill middleware
```
Comment (Gudahtt) [inline app/scripts/controllers/network/createJsonRpcClient.js:25]: Nit: was this comment added in the wrong place? Was there a reason it was added here but not in the Infura client?

**Review by Gudahtt — DISMISSED** [review]
LGTM!

Though we should manually test that this works (I haven't done that yet). I don't think these methods are covered by any e2e tests or our test dapp, unfortunately.

**Comment by Gudahtt** [comment]
We should come up with test instructions so that QA has something to go by.

**Review by mcmire — DISMISSED** [review]
Anything that we add to the network client files we need to make sure is covered in the network client integration tests.

I can work on this.

**Comment by mcmire** [comment]
@Gudahtt Sure, as long as we don't forget about it.

**Review by mcmire — APPROVED** [review]

**Comment by Gudahtt** [comment]
@mcmire I added a ticket here: https://github.com/MetaMask/metamask-extension/issues/16899

**Comment by brad-decker** [comment]
@BelfordZ is this ready to be merged?

**Comment by mcmire** [comment]
I was asked to help with manually testing this PR, and to do this, I've been adding some more buttons and things to the test dapp so we can test the filter and subscription middleware. However, I don't think all of the filter middleware works as intended. I've added a couple of issues around this: https://github.com/MetaMask/eth-json-rpc-filters/issues/80, https://github.com/MetaMask/eth-json-rpc-filters/issues/81. So I'm not able to test all of the functionality but only some of it. I haven't tested the subscription middleware yet, but I will do that tomorrow.

**Comment by mcmire** [comment]
Here are the changes for the test dapp which allow us to test filters and subscriptions: https://github.com/MetaMask/test-dapp/pull/202

I used this on this branch and it seems that there are some issues:

* When creating a log filter, I get this JSON-RPC response: `{"jsonrpc":"2.0","error":{"code":-32600,"message":"invalid json request"}}`
* Creating a block filter works just fine
* When creating a subscription (either way), I get this response: `The method "eth_subscribe" does not exist / is not available.`

**Review by mcmire — DISMISSED** [review]
Marking this as changes requested. I can look into this tomorrow.

File: app/scripts/metamask-controller.js [hunk 2]
```diff
@@ -3920,8 +3906,6 @@ export default class MetamaskController extends EventEmitter {
     ///: END:ONLY_INCLUDE_IN
 
     // filter and subscription polyfills
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:3908]: This comment was regarding the removed lines - it should also be removed.
```suggestion
```

File: app/scripts/controllers/network/createInfuraClient.js [hunk 3]
```diff
@@ -31,8 +42,10 @@ export default function createInfuraClient({ network, projectId }) {
     createRetryOnEmptyMiddleware({ blockTracker, provider: infuraProvider }),
     createBlockTrackerInspectorMiddleware({ blockTracker }),
     infuraMiddleware,
+    filterMiddleware,
```
Comment (Gudahtt) [inline app/scripts/controllers/network/createInfuraClient.js:45]: Oh no! I just realized, the order has changed here. These should be first in the list, but they're last. That probably explains the errors.

Comment (mcmire) [reply app/scripts/controllers/network/createInfuraClient.js:45]: Ah :)

File: app/scripts/controllers/network/createInfuraClient.js [hunk 4]
```diff
@@ -23,6 +25,15 @@ export default function createInfuraClient({ network, projectId }) {
   const infuraProvider = providerFromMiddleware(infuraMiddleware);
   const blockTracker = new PollingBlockTracker({ provider: infuraProvider });
 
+  const filterMiddleware = createFilterMiddleware({
+    infuraProvider,
+    blockTracker,
+  });
+  const subscriptionManager = createSubscriptionManager({
+    infuraProvider,
```
Comment (Gudahtt) [inline app/scripts/controllers/network/createInfuraClient.js:33]: The provider parameters have the wrong names here:
```suggestion
    provider: infuraProvider,
    blockTracker,
  });
  const subscriptionManager = createSubscriptionManager({
    provider: infuraProvider,
```

File: app/scripts/controllers/network/createJsonRpcClient.js [hunk 5]
```diff
@@ -19,11 +21,22 @@ const getTestMiddlewares = () => {
...
     ...blockTrackerOpts,
     provider: blockProvider,
   });
 
+  const filterMiddleware = createFilterMiddleware({
+    blockProvider,
+    blockTracker,
+  });
+  const subscriptionManager = createSubscriptionManager({
+    blockProvider,
```
Comment (Gudahtt) [inline app/scripts/controllers/network/createJsonRpcClient.js:36]: The provider parameters have the wrong names here:

```suggestion
    provider: blockProvider,
    blockTracker,
  });
  const subscriptionManager = createSubscriptionManager({
    provider: blockProvider,
```

**Review by Gudahtt — COMMENTED** [review]
It seems that the subscription middleware is failing because it used to be _before_ the permission middleware, but now it's _after_, and the subscription methods are missing from the list of unrestricted methods (in `app/scripts/controllers/permissions/specifications.js`, the `unrestrictedMethods` variable).

**Comment by Gudahtt** [comment]
In testing out this PR, even with the suggestions I made above applied locally, I'm still having trouble getting subscriptions to work. It looks like there are two distinct problems: notifications are not propagating correctly from the network client to the dapp, and we're seeing occasional empty block responses from Infura that are resulting in the request failing.

For the empty block responses problem, this is happening due to a race condition between `getLatestBlock` and `getBlockByNumber` - there's a period of time during which the latest block number cannot be retrieved by `getBlockByNumber`,it lags behind slightly. This is what the `retryOnEmpty` middleware was doing for us - it would automatically retry.

We'll need to move these middleware up a level, into the network controller, so that we can pass them a `provider` that includes the `retryOnEmpty` middleware.

File: app/scripts/metamask-controller.js [hunk 6]
```diff
@@ -3920,8 +3906,6 @@ export default class MetamaskController extends EventEmitter {
     ///: END:ONLY_INCLUDE_IN
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:3907]: I was able to fix the subscriptions by forwarding logs from the provider to this provider engine, like this:

```suggestion

    // forward notifications from network provider
    provider.on('data', (error, message) => {
      if (error) {
        // This should never happen, this error parameter is never set
        throw error;
      }
      engine.emit('notification', message);
    });
```

Comment (mcmire) [reply app/scripts/metamask-controller.js:3917]: Moved this up a bit but added here: [469bbed](https://github.com/MetaMask/metamask-extension/pull/16693/commits/469bbed580be40d466cc76a97d7910f727eea5cb)

File: app/scripts/controllers/network/createJsonRpcClient.js (hunk 5 above)
Comment (mcmire) [reply app/scripts/controllers/network/createJsonRpcClient.js]: Corrected here: [469bbed](https://github.com/MetaMask/metamask-extension/pull/16693/commits/469bbed580be40d466cc76a97d7910f727eea5cb)

File: app/scripts/controllers/network/createInfuraClient.js (hunk 4 above)
Comment (mcmire) [reply app/scripts/controllers/network/createInfuraClient.js]: Corrected here: [469bbed](https://github.com/MetaMask/metamask-extension/pull/16693/commits/469bbed580be40d466cc76a97d7910f727eea5cb)

File: app/scripts/controllers/network/createInfuraClient.js (hunk 3 above)
Comment (mcmire) [reply app/scripts/controllers/network/createInfuraClient.js]: Made sure to put these first here: [469bbed](https://github.com/MetaMask/metamask-extension/pull/16693/commits/469bbed580be40d466cc76a97d7910f727eea5cb)

File: app/scripts/metamask-controller.js (hunk 2 above)
Comment (mcmire) [reply app/scripts/metamask-controller.js]: Removed here: [e51455a](https://github.com/MetaMask/metamask-extension/pull/16693/commits/e51455a36612ef7a1feb40880863e74c77b0b7f5)

File: app/scripts/controllers/network/createJsonRpcClient.js (hunk 1 above)
Comment (mcmire) [reply app/scripts/controllers/network/createJsonRpcClient.js]: Removed here: [e51455a](https://github.com/MetaMask/metamask-extension/pull/16693/commits/e51455a36612ef7a1feb40880863e74c77b0b7f5)

File: app/scripts/controllers/network/network-controller.js [hunk 7]
```diff
@@ -447,13 +452,30 @@ export default class NetworkController extends EventEmitter {
   }
 
   _setNetworkClient({ networkMiddleware, blockTracker }) {
+    const networkProvider = providerFromMiddleware(networkMiddleware);
```
Comment (mcmire) [inline app/scripts/controllers/network/network-controller.js:455]: @Gudahtt Is this what you were looking for? After this I am not sure if I would be able to tell the difference between the "MetaMask middleware" and the "network middleware", but I guess this feels better...

Comment (Gudahtt) [reply app/scripts/controllers/network/network-controller.js:455]: Yeah this is exactly what I had in mind!

The "metamask middleware" is being removed in this PR: https://github.com/MetaMask/metamask-extension/pull/16863 . It won't be here long.

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by brad-decker — APPROVED** [review]

**Review by jpuri — APPROVED** [review]

**Pull request merged**